docs/tutorials/data_provider.md

```diff
@@ -61,3 +61,169 @@ impl AdditiveIdentity {
     }
 }
 ```
+
+## Caching Data Provider
+
+ICU4X has no internal caches because there is no one-size-fits-all solution. It is easy for clients to implement their own cache for ICU4X, and although this is not generally required or recommended, it may be beneficial when latency is of utmost importance and, for example, a less-efficient data provider such as JSON is being used.
+
+The following example illustrates an LRU cache on top of a BufferProvider that saves deserialized data payloads as type-erased objects and then checks for a cache hit before calling the inner provider.
+
+```rust
+use icu_provider::hello_world::HelloWorldFormatter;
+use icu_provider::prelude::*;
+use icu::locid::locale;
+use lru::LruCache;
+use std::borrow::{Borrow, Cow};
+use std::convert::TryInto;
+use std::sync::Mutex;
+use yoke::trait_hack::YokeTraitHack;
+use yoke::Yokeable;
+use zerofrom::ZeroFrom;
+
+#[derive(Debug, PartialEq, Eq, Hash)]
+struct CacheKeyWrap(CacheKey<'static>);
+
+#[derive(Debug, PartialEq, Eq, Hash)]
+struct CacheKey<'a>(DataKey, Cow<'a, DataLocale>);
```

Comment:
suggestion: comment explanation on this section about how the borrow works here to make it possible to do non-cloning gets (probably pair CacheKey with the Borrow impl in code organization too)

Reply:
Yeah. I'm not proud of this code because it uses a `#[doc(hidden)]` export, `lru::KeyRef`. It's confusing and distracting to have it in the example. I originally skipped this and just cloned the key for lookup, but @robertbastian's review feedback was that my code was cloning too much. I can labor over explaining how the key borrowing works, but it's not the point of this example.

Reply:
Yeah I think we shouldn't explain it in depth but every line of code in an example is something the reader may focus on: I think we should move the Borrow impl and the borrowed key type below the main stuff with a one-liner comment that just says it's to avoid clones on lookup , so that the reader doesn't spend too much time trying to figure out how it fits in to the rest